onnx/defs/shape_inference.h

```diff
@@ -74,11 +74,27 @@ struct InferenceContext {
   virtual const SparseTensorProto* getInputSparseData(size_t index) const = 0;
 };
 
+struct DataPropagationContext {
+  virtual const AttributeProto* getAttribute(const std::string& name) const = 0;
+  virtual size_t getNumInputs() const = 0;
+  virtual const TypeProto* getInputType(size_t index) const = 0;
+  virtual size_t getNumOutputs() const = 0;
+  virtual TypeProto* getOutputType(size_t index) = 0;
+  virtual ~DataPropagationContext() {}
+  virtual const TensorProto* getInputData(size_t index) const = 0;
+  virtual void addOutputShapeData(size_t index, TensorShapeProto&& tp) = 0;
+  virtual const TensorShapeProto* getInputShapeData(size_t index) const = 0;
```

Comment:
Instead of having 
virtual const TensorProto* getInputData(size_t index) const = 0;
 virtual const TensorShapeProto* getInputShapeData(size_t index) const = 0;

can you simply keep one like this:
virtual const TensorShapeProto* getInputData(size_t index) const = 0;

The implementation of DataPropagationContext can convert the TensorProto to TensorShapeProto... otherwise in every data propagation function you will need to call 1 and if it returns nullptr then call the other...

Reply:
Merging it as single `virtual const TensorShapeProto* getInputData(size_t index) const = 0;` sounds good. The updated getInputData will first check whether `allInputTypes_` exists a valid Shape inside TypeProto. If not, it will further check TensorShapeProto from `generatedShapeData_`. Please review it again